In [1]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [2]:
import pandas as pd
filename = '/content/drive/MyDrive/Colab Notebooks/02. Coding Dojo/Semana 03/01. Plataforma/Proyecto 01 - Parte 02/sales_predictions.csv'
df = pd.read_csv(filename)
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [3]:
df.isnull().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Analizaremos los valores que tengan nulos: 'Item Weight" y "Oulet_Size"

In [4]:
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

Se tendrá que hacer una corrección más adelante para estandarizar esta columna, sin embargo solo se presentan dos tipos. Low Fat y Regular. Existen 1463 valores nuloes para la columna Item_Weight.

In [5]:
#Doble Filtro para saber si hay filas con valores faltantes de Item_Weight y Outlet_Size
filter_weight = df['Item_Weight'].isnull()
filter_outlet_size = df['Outlet_Size'].isnull()

df.loc[filter_weight & filter_outlet_size,:]

Empty DataFrame
Columns: [Item_Identifier, Item_Weight, Item_Fat_Content, Item_Visibility, Item_Type, Item_MRP, Outlet_Identifier, Outlet_Establishment_Year, Outlet_Size, Outlet_Location_Type, Outlet_Type, Item_Outlet_Sales]
Index: []

*Observamos de que no hay filas faltantes las cuales tengan valores nulos tanto en peso y talla, por lo que trataremos de encontrar una relación para reemplazar los datos faltantes.*

### a) Recuperación de los valores de Outlet_size

In [6]:
df['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [7]:
df1 = df.loc[df['Outlet_Size'].isnull(),:]
df1

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3              FDX07       19.200          Regular         0.000000   
8              FDH17       16.200          Regular         0.016687   
9              FDU28       19.200          Regular         0.094450   
25             NCD06       13.000          Low Fat         0.099887   
28             FDE51        5.925          Regular         0.161467   
...              ...          ...              ...              ...   
8502           NCH43        8.420          Low Fat         0.070712   
8508           FDW31       11.350          Regular         0.043246   
8509           FDG45        8.100          Low Fat         0.214306   
8514           FDA01       15.000          Regular         0.054489   
8519           FDS36        8.380          Regular         0.046982   

                  Item_Type  Item_MRP Outlet_Identifier  \
3     Fruits and Vegetables  182.0950            OUT010   
8              Frozen Foods   96.9726            OUT045   
9              Frozen Foods  187.8214            OUT017   
25                Household   45.9060            OUT017   
28                    Dairy   45.5086            OUT010   
...                     ...       ...               ...   
8502              Household  216.4192            OUT045   
8508  Fruits and Vegetables  199.4742            OUT045   
8509  Fruits and Vegetables  213.9902            OUT010   
8514                 Canned   57.5904            OUT045   
8519           Baking Goods  108.1570            OUT045   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
3                          1998         NaN               Tier 3   
8                          2002         NaN               Tier 2   
9                          2007         NaN               Tier 2   
25                         2007         NaN               Tier 2   
28                         1998         NaN               Tier 3   
...                         ...         ...                  ...   
8502                       2002         NaN               Tier 2   
8508                       2002         NaN               Tier 2   
8509                       1998         NaN               Tier 3   
8514                       2002         NaN               Tier 2   
8519                       2002         NaN               Tier 2   

            Outlet_Type  Item_Outlet_Sales  
3         Grocery Store           732.3800  
8     Supermarket Type1          1076.5986  
9     Supermarket Type1          4710.5350  
25    Supermarket Type1           838.9080  
28        Grocery Store           178.4344  
...                 ...                ...  
8502  Supermarket Type1          3020.0688  
8508  Supermarket Type1          2587.9646  
8509      Grocery Store           424.7804  
8514  Supermarket Type1           468.7232  
8519  Supermarket Type1           549.2850  

[2410 rows x 12 columns]

Se identifica una frecuencia en los valores de Outlet_Identifier. Además de que están relacionados los valores de Outlet_Identifier con Locationtype y oulet type

In [8]:
df1['Outlet_Identifier'].value_counts()

OUT045    929
OUT017    926
OUT010    555
Name: Outlet_Identifier, dtype: int64

Regresamos a nuestro df original para ver si existe una relación macro entre el identificador y el tipo de locación.

In [9]:
df.groupby(['Outlet_Type', 'Outlet_Location_Type', 'Outlet_Identifier'])[['Item_Identifier']].count()

Item_Identifier
Outlet_Type       Outlet_Location_Type Outlet_Identifier                 
Grocery Store     Tier 1               OUT019                         528
                  Tier 3               OUT010                         555
Supermarket Type1 Tier 1               OUT046                         930
                                       OUT049                         930
                  Tier 2               OUT017                         926
                                       OUT035                         930
                                       OUT045                         929
                  Tier 3               OUT013                         932
Supermarket Type2 Tier 3               OUT018                         928
Supermarket Type3 Tier 3               OUT027                         935

De esto observamos que todos los valores pertenencen a una misma tienda.

- OUT017 > Supermarket Type1
- OUT045 > Supermarket Type1
- OUT010 > Grocery Store

Ahora compararemos las ventas totales entre cada tipo de tienda

In [10]:
df.groupby(['Outlet_Type','Outlet_Location_Type','Outlet_Identifier','Outlet_Size'])[['Item_Outlet_Sales']].sum().round(2)

Item_Outlet_Sales
Outlet_Type       Outlet_Location_Type Outlet_Identifier Outlet_Size                   
Grocery Store     Tier 1               OUT019            Small                179694.09
Supermarket Type1 Tier 1               OUT046            Small               2118395.17
                                       OUT049            Medium              2183969.81
                  Tier 2               OUT035            Small               2268122.94
                  Tier 3               OUT013            High                2142663.58
Supermarket Type2 Tier 3               OUT018            Medium              1851822.83
Supermarket Type3 Tier 3               OUT027            Medium              3453926.05

No salen los identificadores completos, ya que se tiene valores faltantes dentro de los Identificadores, quitaremos outlet Size para ver alguna relación

In [11]:
df.groupby(['Outlet_Type','Outlet_Location_Type','Outlet_Identifier'])[['Item_Outlet_Sales']].sum().round(2)

Item_Outlet_Sales
Outlet_Type       Outlet_Location_Type Outlet_Identifier                   
Grocery Store     Tier 1               OUT019                     179694.09
                  Tier 3               OUT010                     188340.17
Supermarket Type1 Tier 1               OUT046                    2118395.17
                                       OUT049                    2183969.81
                  Tier 2               OUT017                    2167465.29
                                       OUT035                    2268122.94
                                       OUT045                    2036725.48
                  Tier 3               OUT013                    2142663.58
Supermarket Type2 Tier 3               OUT018                    1851822.83
Supermarket Type3 Tier 3               OUT027                    3453926.05

- Visualizamos que dentro de Grocery Store, OUT010 Tiene mayores salidas de ventas que OUT019 (Small). Por lo que se le asignará un (Medium).
- Por otro lado en la categoría Supermercado tipo 1 y tier 2. Tenemos:

OUT017 >> 2.167
OUT035 >> 2.268 (Small)
OUT045 >> 2.036

Cuyos ambos valores son inferiores, por lo que se procederá a asignarles un valor de Small

Quedando:

- OUT017 > Small
- OUT045 > Small
- OUT010 > Medium

In [12]:
# Reemplazamos los valores correspondientes
OUT010_filter = df['Outlet_Identifier'] == 'OUT010'
OUT045_filter = df['Outlet_Identifier'] == 'OUT045'
OUT017_filter = df['Outlet_Identifier'] == 'OUT017'

df.loc[OUT010_filter, 'Outlet_Size']=df.loc[OUT010_filter, 'Outlet_Size'].fillna('Medium')
df.loc[(OUT017_filter | OUT045_filter), 'Outlet_Size']=df.loc[(OUT017_filter | OUT045_filter), 'Outlet_Size'].fillna('Small')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                8523 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [13]:
df.groupby(['Outlet_Type','Outlet_Location_Type','Outlet_Identifier','Outlet_Size'])[['Item_Outlet_Sales']].sum().round(2)

Item_Outlet_Sales
Outlet_Type       Outlet_Location_Type Outlet_Identifier Outlet_Size                   
Grocery Store     Tier 1               OUT019            Small                179694.09
                  Tier 3               OUT010            Medium               188340.17
Supermarket Type1 Tier 1               OUT046            Small               2118395.17
                                       OUT049            Medium              2183969.81
                  Tier 2               OUT017            Small               2167465.29
                                       OUT035            Small               2268122.94
                                       OUT045            Small               2036725.48
                  Tier 3               OUT013            High                2142663.58
Supermarket Type2 Tier 3               OUT018            Medium              1851822.83
Supermarket Type3 Tier 3               OUT027            Medium              3453926.05

### b) Identificar valores de Item_Fat_Content

In [14]:
df

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDA15        9.300          Low Fat         0.016047   
1              DRC01        5.920          Regular         0.019278   
2              FDN15       17.500          Low Fat         0.016760   
3              FDX07       19.200          Regular         0.000000   
4              NCD19        8.930          Low Fat         0.000000   
...              ...          ...              ...              ...   
8518           FDF22        6.865          Low Fat         0.056783   
8519           FDS36        8.380          Regular         0.046982   
8520           NCJ29       10.600          Low Fat         0.035186   
8521           FDN46        7.210          Regular         0.145221   
8522           DRG01       14.800          Low Fat         0.044878   

                  Item_Type  Item_MRP Outlet_Identifier  \
0                     Dairy  249.8092            OUT049   
1               Soft Drinks   48.2692            OUT018   
2                      Meat  141.6180            OUT049   
3     Fruits and Vegetables  182.0950            OUT010   
4                 Household   53.8614            OUT013   
...                     ...       ...               ...   
8518            Snack Foods  214.5218            OUT013   
8519           Baking Goods  108.1570            OUT045   
8520     Health and Hygiene   85.1224            OUT035   
8521            Snack Foods  103.1332            OUT018   
8522            Soft Drinks   75.4670            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                          1999      Medium               Tier 1   
1                          2009      Medium               Tier 3   
2                          1999      Medium               Tier 1   
3                          1998      Medium               Tier 3   
4                          1987        High               Tier 3   
...                         ...         ...                  ...   
8518                       1987        High               Tier 3   
8519                       2002       Small               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
0     Supermarket Type1          3735.1380  
1     Supermarket Type2           443.4228  
2     Supermarket Type1          2097.2700  
3         Grocery Store           732.3800  
4     Supermarket Type1           994.7052  
...                 ...                ...  
8518  Supermarket Type1          2778.3834  
8519  Supermarket Type1           549.2850  
8520  Supermarket Type1          1193.1136  
8521  Supermarket Type2          1845.5976  
8522  Supermarket Type1           765.6700  

[8523 rows x 12 columns]

In [15]:
df.isnull().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                     0
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [16]:
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

Primero vamos a normalizar los valores

In [17]:
df["Item_Fat_Content"].replace("LF", "Low Fat", inplace=True)
df["Item_Fat_Content"].replace("low fat", "Low Fat", inplace=True)
df["Item_Fat_Content"].replace("reg", "Regular", inplace=True)
df["Item_Fat_Content"].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

Partiendo de la hipotesis que otras tiendas vendan los mismo productos, buscaremos una forma de relacionarlos. Basandonos del ID unico del producto

Item_Identifier	Número de identificación único del producto

In [18]:
df['Item_Identifier'].value_counts().reset_index()

index  Item_Identifier
0     FDW13               10
1     FDG33               10
2     NCY18                9
3     FDD38                9
4     DRE49                9
...     ...              ...
1554  FDY43                1
1555  FDQ60                1
1556  FDO33                1
1557  DRF48                1
1558  FDC23                1

[1559 rows x 2 columns]

Existen 1559 valores únicos registrados. Ahora analizaremos los valores los cuales no tienen peso

In [48]:
df.groupby(['Item_Identifier', 'Item_Weight'])[['Item_Weight']].count()

Item_Weight
Item_Identifier Item_Weight             
DRA12           11.600                 6
DRA24           19.350                 5
DRA59           8.270                  6
DRB01           7.390                  2
DRB13           6.115                  5
...                                  ...
NCZ30           6.590                  6
NCZ41           19.850                 5
NCZ42           10.500                 5
NCZ53           9.600                  4
NCZ54           14.650                 5

[1555 rows x 1 columns]

Existe una diferencia de 4 filas, lo que significa que hay cuatro valores únicos los cuales no registraron peso.

In [20]:
# Elaboramos una lista de productos únicos
lisn1=[]
for x in df['Item_Identifier'].unique():
  lisn1.append(x)
print(lisn1)

['FDA15', 'DRC01', 'FDN15', 'FDX07', 'NCD19', 'FDP36', 'FDO10', 'FDP10', 'FDH17', 'FDU28', 'FDY07', 'FDA03', 'FDX32', 'FDS46', 'FDF32', 'FDP49', 'NCB42', 'DRI11', 'FDU02', 'FDN22', 'FDW12', 'NCB30', 'FDC37', 'FDR28', 'NCD06', 'FDV10', 'DRJ59', 'FDE51', 'FDC14', 'FDV38', 'NCS17', 'FDP33', 'FDO23', 'DRH01', 'NCX29', 'FDV20', 'DRZ11', 'FDX10', 'FDB34', 'FDK43', 'FDA46', 'FDC02', 'FDL50', 'FDM39', 'NCP05', 'FDV49', 'FDL12', 'FDS02', 'NCL17', 'FDM40', 'FDR13', 'FDA43', 'NCP18', 'FDK21', 'NCX54', 'DRK35', 'FDY21', 'FDI26', 'FDM20', 'FDV27', 'FDF09', 'FDY40', 'FDY45', 'FDC46', 'FDH19', 'FDZ03', 'DRH37', 'NCI17', 'FDJ58', 'FDH35', 'FDG02', 'NCZ18', 'FDC29', 'FDQ10', 'FDN48', 'FDL04', 'FDV25', 'FDD58', 'FDN04', 'FDV45', 'NCL18', 'FDR12', 'FDG20', 'FDZ55', 'FDQ49', 'FDN33', 'FDN27', 'FDW20', 'DRG27', 'DRI25', 'FDA44', 'NCR17', 'FDU04', 'FDF41', 'FDB56', 'FDT28', 'FDD10', 'FDW57', 'DRB48', 'FDP09', 'FDH14', 'FDA47', 'FDG12', 'DRE60', 'DRK49', 'FDD03', 'FDS52', 'FDW39', 'FDX34', 'FDV11', 'FDD17', 

In [50]:
df_muestra = df.groupby(['Item_Identifier', 'Item_Weight'])[['Item_Visibility']].count().reset_index()
df_muestra

Item_Identifier  Item_Weight  Item_Visibility
0              DRA12       11.600                6
1              DRA24       19.350                5
2              DRA59        8.270                6
3              DRB01        7.390                2
4              DRB13        6.115                5
...              ...          ...              ...
1550           NCZ30        6.590                6
1551           NCZ41       19.850                5
1552           NCZ42       10.500                5
1553           NCZ53        9.600                4
1554           NCZ54       14.650                5

[1555 rows x 3 columns]

In [22]:
print(df_muestra['Item_Identifier'])

0       DRA12
1       DRA24
2       DRA59
3       DRB01
4       DRB13
        ...  
1550    NCZ30
1551    NCZ41
1552    NCZ42
1553    NCZ53
1554    NCZ54
Name: Item_Identifier, Length: 1555, dtype: object


In [39]:
set_lista_grande = pd.DataFrame(lisn1, columns=['Item_Identifier'])
set_lista_pequeña = pd.DataFrame(df_muestra['Item_Identifier'])

In [72]:
## Ubicamos valores unicos faltantes
valores_faltantes = set_lista_grande[~set_lista_grande['Item_Identifier'].isin(set_lista_pequeña['Item_Identifier'])]
valores_faltantes

Item_Identifier
711            FDN52
1130           FDK57
1494           FDE52
1539           FDQ60

Vamos uno a uno para identificar si hay valores que se podría completar

In [66]:
#Revisión FDN52
Id_filter=df['Item_Identifier'] == 'FDN52'
df.loc[Id_filter]['Item_Weight']

927   NaN
Name: Item_Weight, dtype: float64

In [67]:
#Revisión FDK57
Id_filter=df['Item_Identifier'] == 'FDK57'
df.loc[Id_filter]['Item_Weight']

1922   NaN
Name: Item_Weight, dtype: float64

In [69]:
#Revisión FDE52
Id_filter=df['Item_Identifier'] == 'FDE52'
df.loc[Id_filter]['Item_Weight']

4187   NaN
Name: Item_Weight, dtype: float64

In [68]:
#Revisión FDQ60
Id_filter=df['Item_Identifier'] == 'FDQ60'
df.loc[Id_filter]['Item_Weight']

5022   NaN
Name: Item_Weight, dtype: float64

Vamos con los valores nulos y que tiene identificador

In [74]:
df_null_weight = df.loc[df['Item_Weight'].isnull(),:]
df_null_weight

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
7              FDP10          NaN          Low Fat         0.127470   
18             DRI11          NaN          Low Fat         0.034238   
21             FDW12          NaN          Regular         0.035400   
23             FDC37          NaN          Low Fat         0.057557   
29             FDC14          NaN          Regular         0.072222   
...              ...          ...              ...              ...   
8485           DRK37          NaN          Low Fat         0.043792   
8487           DRG13          NaN          Low Fat         0.037006   
8488           NCN14          NaN          Low Fat         0.091473   
8490           FDU44          NaN          Regular         0.102296   
8504           NCN18          NaN          Low Fat         0.124111   

                  Item_Type  Item_MRP Outlet_Identifier  \
7               Snack Foods  107.7622            OUT027   
18              Hard Drinks  113.2834            OUT027   
21             Baking Goods  144.5444            OUT027   
23             Baking Goods  107.6938            OUT019   
29                   Canned   43.6454            OUT019   
...                     ...       ...               ...   
8485            Soft Drinks  189.0530            OUT027   
8487            Soft Drinks  164.7526            OUT027   
8488                 Others  184.6608            OUT027   
8490  Fruits and Vegetables  162.3552            OUT019   
8504              Household  111.7544            OUT027   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
7                          1985      Medium               Tier 3   
18                         1985      Medium               Tier 3   
21                         1985      Medium               Tier 3   
23                         1985       Small               Tier 1   
29                         1985       Small               Tier 1   
...                         ...         ...                  ...   
8485                       1985      Medium               Tier 3   
8487                       1985      Medium               Tier 3   
8488                       1985      Medium               Tier 3   
8490                       1985       Small               Tier 1   
8504                       1985      Medium               Tier 3   

            Outlet_Type  Item_Outlet_Sales  
7     Supermarket Type3          4022.7636  
18    Supermarket Type3          2303.6680  
21    Supermarket Type3          4064.0432  
23        Grocery Store           214.3876  
29        Grocery Store           125.8362  
...                 ...                ...  
8485  Supermarket Type3          6261.8490  
8487  Supermarket Type3          4111.3150  
8488  Supermarket Type3          2756.4120  
8490      Grocery Store           487.3656  
8504  Supermarket Type3          4138.6128  

[1463 rows x 12 columns]

In [89]:
df_list_null = df_null_weight['Item_Identifier'].value_counts().reset_index()
df_list_null = df_list_null.rename(columns={'index': 'Item_Identifier','Item_Identifier': 'Item_Identifier_count'})
df_list_null

Item_Identifier  Item_Identifier_count
0              FDK08                      2
1              FDA08                      2
2              FDV23                      2
3              FDY56                      2
4              FDI04                      2
...              ...                    ...
1137           FDM44                      1
1138           FDZ48                      1
1139           FDK41                      1
1140           FDD57                      1
1141           NCN18                      1

[1142 rows x 2 columns]

In [94]:
df_base = df.groupby(['Item_Identifier', 'Item_Weight'])[['Item_Visibility']].count().reset_index()
df_peso_identifier = df_base[['Item_Identifier','Item_Weight']]
df_peso_identifier

Item_Identifier  Item_Weight
0              DRA12       11.600
1              DRA24       19.350
2              DRA59        8.270
3              DRB01        7.390
4              DRB13        6.115
...              ...          ...
1550           NCZ30        6.590
1551           NCZ41       19.850
1552           NCZ42       10.500
1553           NCZ53        9.600
1554           NCZ54       14.650

[1555 rows x 2 columns]

Buscamos Item a Item, valores repetidos y los asignamos a los valores con un valor null.

In [95]:
for x in lisn1:
  id_filter=df['Item_Identifier'] == x
  m=df.loc[id_filter]['Item_Weight'].max()
  df.loc[id_filter,'Item_Weight']=df.loc[id_filter]['Item_Weight'].fillna(m)

In [98]:
#validamos duplicados
df.duplicated().sum()

0

In [96]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                8519 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                8523 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Corroboramos que solo existen 4 valores los cuales no fueron reemplazados, esto debido a que son valores únicos.

In [97]:
tabla_null=df[df['Item_Weight'].isna()]
tabla_null

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
927            FDN52          NaN          Regular         0.130933   
1922           FDK57          NaN          Low Fat         0.079904   
4187           FDE52          NaN          Regular         0.029742   
5022           FDQ60          NaN          Regular         0.191501   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
927   Frozen Foods   86.9198            OUT027                       1985   
1922   Snack Foods  120.0440            OUT027                       1985   
4187         Dairy   88.9514            OUT027                       1985   
5022  Baking Goods  121.2098            OUT019                       1985   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
927       Medium               Tier 3  Supermarket Type3          1569.9564  
1922      Medium               Tier 3  Supermarket Type3          4434.2280  
4187      Medium               Tier 3  Supermarket Type3          3453.5046  
5022       Small               Tier 1      Grocery Store           120.5098

In [99]:
#Identificamos a qué tipo pertenecen
lisn2=list(tabla_null['Item_Type'])
lisn2

['Frozen Foods', 'Snack Foods', 'Dairy', 'Baking Goods']

Para no perder estos datos, ya que los Item Oulet sales es un valor considerable, haremos el promedio de los item type y les asignaremos ese valor.

In [100]:
for x in lisn2:
  type_filter=df['Item_Type'] == x
  mean=df.loc[type_filter]['Item_Weight'].mean()
  df.loc[type_filter,'Item_Weight']=df.loc[type_filter]['Item_Weight'].fillna(mean)

Finalmente verificamos que no existan valores nulos

In [101]:
df[df['Item_Weight'].isna()]

Empty DataFrame
Columns: [Item_Identifier, Item_Weight, Item_Fat_Content, Item_Visibility, Item_Type, Item_MRP, Outlet_Identifier, Outlet_Establishment_Year, Outlet_Size, Outlet_Location_Type, Outlet_Type, Item_Outlet_Sales]
Index: []

In [105]:
#Y verificamos valores estadísticos de algunas columnas
df.describe().T

count         mean          std       min  \
Item_Weight                8523.0    12.875418     4.645016     4.555   
Item_Visibility            8523.0     0.066132     0.051598     0.000   
Item_MRP                   8523.0   140.992782    62.275067    31.290   
Outlet_Establishment_Year  8523.0  1997.831867     8.371760  1985.000   
Item_Outlet_Sales          8523.0  2181.288914  1706.499616    33.290   

                                   25%          50%          75%           max  
Item_Weight                   8.785000    12.650000    16.850000     21.350000  
Item_Visibility               0.026989     0.053931     0.094585      0.328391  
Item_MRP                     93.826500   143.012800   185.643700    266.888400  
Outlet_Establishment_Year  1987.000000  1999.000000  2004.000000   2009.000000  
Item_Outlet_Sales           834.247400  1794.331000  3101.296400  13086.964800